File: src/Entity/Excursion.php
<?php

namespace App\Entity;

use App\Repository\ExcursionRepository;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use App\Entity\Rating;

#[ORM\Entity(repositoryClass: ExcursionRepository::class)]
#[ORM\Table(name: 'excursion')]
class Excursion
{
    // ===================== ID =====================
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'idExcursion', type: 'integer')]
    // @phpstan-ignore-next-line
    private ?int $idExcursion = null;

    // ===================== RUNTIME-ONLY (not persisted) =====================
    /** @var mixed */
    public mixed $prixData = null;

    /** @var array<string, mixed>|null */
    public ?array $weather = null;

    // ===================== INFOS PRINCIPALES =====================
    #[ORM\Column(length: 150, nullable: false)]
    #[Assert\NotBlank(message: "Le titre est obligatoire")]
    #[Assert\Length(min: 3)]
    #[Assert\Regex(pattern: "/^[^<>]*$/", message: "HTML interdit")]
    private string $titre = '';  // ✅

    #[ORM\Column(type: Types::TEXT, nullable: false)]
    #[Assert\NotBlank]
    #[Assert\Length(min: 15)]
    #[Assert\Regex(pattern: "/^[^<>]*$/", message: "HTML interdit")]
    private string $description = '';  // ✅

    #[ORM\Column(length: 100, nullable: false)]
    #[Assert\NotBlank]
    #[Assert\Length(min: 3)]
    #[Assert\Regex(pattern: "/^[^<>]*$/", message: "HTML interdit")]
    private string $lieu = '';  // ✅

    #[ORM\Column(type: Types::DATE_MUTABLE, nullable: false)]
    #[Assert\NotNull]
    #[Assert\GreaterThan("today")]
    private \DateTimeInterface $date;  // ✅

    #[ORM\Column(type: 'integer', nullable: false)]
    #[Assert\NotBlank]
    #[Assert\Positive]
    private int $duree = 0;  // ✅

    // ===================== PLACES =====================
    #[ORM\Column(name: 'nombrePlacesDisponibles', type: 'integer', nullable: false)]
    #[Assert\NotBlank]
    #[Assert\Positive]
    private int $nombrePlacesDisponibles = 0;  // ✅

    // ===================== PRIX =====================
    #[ORM\Column(name: 'prix_adulte', type: 'float', nullable: false)]
    #[Assert\NotBlank]
    #[Assert\GreaterThanOrEqual(0)]
    private float $prixAdulte = 0.0;  // ✅

    #[ORM\Column(name: 'prix_enfant', type: 'float', nullable: false)]
    #[Assert\NotBlank]
    #[Assert\GreaterThanOrEqual(0)]
    private float $prixEnfant = 0.0;  // ✅

    #[ORM\Column(name: 'prix_bebe', type: 'float', nullable: false)]
    #[Assert\NotBlank]
    #[Assert\GreaterThanOrEqual(0)]
    private float $prixBebe = 0.0;  // ✅

    // ===================== AUTRES =====================
    #[ORM\Column(name: 'image_url', length: 500, nullable: true)]
    private ?string $imageUrl = null;

    #[ORM\Column(length: 20, nullable: true, options: ['default' => 'OUTDOOR'])]
    private ?string $type = 'OUTDOOR';

    // ===================== RELATIONS =====================
    #[ORM\ManyToOne(targetEntity: Categorie::class, inversedBy: 'excursions')]
    #[ORM\JoinColumn(nullable: false)]
    #[Assert\NotNull]
    private ?Categorie $categorie = null;

    /** @var Collection<int, Rating> */
    #[ORM\OneToMany(mappedBy: 'excursion', targetEntity: Rating::class)]
    private Collection $ratings;

    // ===================== MAP =====================
    #[ORM\Column(type: 'float', nullable: true)]
    private ?float $latitude = null;

    #[ORM\Column(type: 'float', nullable: true)]
    private ?float $longitude = null;

    // ===================== CONSTRUCT =====================
    public function __construct()
    {
        $this->ratings = new ArrayCollection();
    }

    // ===================== GETTERS =====================
    public function getIdExcursion(): ?int { return $this->idExcursion; }
    public function getTitre(): string { return $this->titre; }
    public function getDescription(): string { return $this->description; }
    public function getLieu(): string { return $this->lieu; }
    public function getDate(): \DateTimeInterface { return $this->date; }
    public function getDuree(): int { return $this->duree; }
    public function getNombrePlacesDisponibles(): int { return $this->nombrePlacesDisponibles; }
    public function getPrixAdulte(): float { return $this->prixAdulte; }
    public function getPrixEnfant(): float { return $this->prixEnfant; }
    public function getPrixBebe(): float { return $this->prixBebe; }
    public function getImageUrl(): ?string { return $this->imageUrl; }
    public function getType(): ?string { return $this->type; }
    public function getCategorie(): ?Categorie { return $this->categorie; }
    public function getLatitude(): ?float { return $this->latitude; }
    public function getLongitude(): ?float { return $this->longitude; }

    /** @return Collection<int, Rating> */
    public function getRatings(): Collection { return $this->ratings; }

    // ===================== SETTERS =====================
    public function setTitre(string $v): static { $this->titre = $v; return $this; }
    public function setDescription(string $v): static { $this->description = $v; return $this; }
    public function setLieu(string $v): static { $this->lieu = $v; return $this; }
    public function setDate(\DateTimeInterface $v): static { $this->date = $v; return $this; }
    public function setDuree(int $v): static { $this->duree = $v; return $this; }
    public function setNombrePlacesDisponibles(int $v): static { $this->nombrePlacesDisponibles = $v; return $this; }
    public function setPrixAdulte(float $v): static { $this->prixAdulte = $v; return $this; }
    public function setPrixEnfant(float $v): static { $this->prixEnfant = $v; return $this; }
    public function setPrixBebe(float $v): static { $this->prixBebe = $v; return $this; }
    public function setImageUrl(?string $v): static { $this->imageUrl = $v; return $this; }
    public function setType(?string $v): static { $this->type = $v; return $this; }
    public function setCategorie(?Categorie $v): static { $this->categorie = $v; return $this; }
    public function setLatitude(?float $v): static { $this->latitude = $v; return $this; }
    public function setLongitude(?float $v): static { $this->longitude = $v; return $this; }
}

File: src/main/java/tn/esprit/entities/Excursion.java
package tn.esprit.entities;
import java.time.LocalDate;

public class Excursion {
    private int idExcursion;
    private String titre;
    private String description;
    private String lieu;
    private LocalDate date;
    private int duree ; // en heures
    private double prix;
    private int nombrePlacesDisponibles;
    private String imageUrl;
    private double prixAdulte;
    private double prixEnfant;
    private double prixBebe;



    public Excursion(){}
    public Excursion(int idExcursion,
                     String titre,
                     LocalDate date,
                     String description,
                     String lieu,
                     double prixAdulte,
                     double prixEnfant,
                     double prixBebe,
                     int duree,
                     int nombrePlacesDisponibles,
                     String imageUrl) {

        this.idExcursion = idExcursion;
        this.titre = titre;
        this.date = date;
        this.description = description;
        this.lieu = lieu;

        this.prixAdulte = prixAdulte;
        this.prixEnfant = prixEnfant;
        this.prixBebe = prixBebe;

        this.duree = duree;
        this.nombrePlacesDisponibles = nombrePlacesDisponibles;
        this.imageUrl = imageUrl;
    }


    public String getTitre() {
        return titre;
    }

    public void setTitre(String titre) {
        this.titre = titre;
    }

    public int getIdExcursion() {
        return idExcursion;
    }

    public void setIdExcursion(int idExcursion) {
        this.idExcursion = idExcursion;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getLieu() {
        return lieu;
    }

    public void setLieu(String lieu) {
        this.lieu = lieu;
    }

    public LocalDate getDate() {
        return date;
    }

    public void setDate(LocalDate date) {
        this.date = date;
    }

    public int getDuree() {
        return duree;
    }

    public void setDuree(int duree) {
        this.duree = duree;
    }

    public int getNombrePlacesDisponibles() {
        return nombrePlacesDisponibles;
    }

    public void setNombrePlacesDisponibles(int nombrePlacesDisponibles) {
        this.nombrePlacesDisponibles = nombrePlacesDisponibles;
    }

    public String getImageUrl() {
        return imageUrl;
    }

    public void setImageUrl(String imageUrl) {
        this.imageUrl = imageUrl;
    }

    public double getPrixAdulte() {
        return prixAdulte;

    }
    public void setPrixAdulte(double prixAdulte) {
        this.prixAdulte = prixAdulte;
    }
    public double getPrixEnfant() {
        return prixEnfant;
    }
    public void setPrixEnfant(double prixEnfant) {
        this.prixEnfant = prixEnfant;
    }
    public double getPrixBebe() {
        return prixBebe;
    }
    public void setPrixBebe(double prixBebe) {
        this.prixBebe = prixBebe;

    }


    @Override
    public String toString() {
        return "Excursion{" +
                "idExcursion=" + idExcursion +
                ", titre='" + titre + '\'' +
                ", description='" + description + '\'' +
                ", lieu='" + lieu + '\'' +
                ", date=" + date +
                ", duree=" + duree +
                ", prix=" + prix +
                ", nombrePlacesDisponibles=" + nombrePlacesDisponibles +
                ", imageUrl='" + imageUrl + '\'' +
                ", prixAdulte=" + prixAdulte +
                ", prixEnfant=" + prixEnfant +
                ", prixBebe=" + prixBebe +
                '}';
    }


}
